src/AMAMM.sol

```diff
@@ -2,69 +2,244 @@
-                beforeSwapReturnDelta: false,
-                afterSwapReturnDelta: true, // Allow afterSwap to return a custom delta
-                afterAddLiquidityReturnDelta: false,
-                afterRemoveLiquidityReturnDelta: false
-            });
+        /// -----------------------------------------------------------------------
+        /// Validation
+        /// -----------------------------------------------------------------------
+
+        address msgSender = LibMulticaller.senderOrSigner();
+
+        // ensure bid is valid
+        // - manager can't be zero address
+        // - bid needs to be greater than the next bid by >10%
+        // - deposit needs to cover the rent for K hours
+        // - deposit needs to be a multiple of rent
+        // - payload needs to be valid
+        if (
+            bidder == address(0) || rent <= poolEpochManager[id][_epoch].rent.mulWad(MIN_BID_MULTIPLIER(id))
+                || deposit < rent * K(id) || deposit % rent != 0 || !_payloadIsValid(id, payload)
+        ) {
+            revert AmAmm__InvalidBid();
+        }
+
+        // Check if the bid already exists
+        Bid memory existingBid = poolEpochBids[id][_epoch][bidder];
+        Bid memory newBid = Bid({bidder: msgSender, payload: payload, rent: rent, deposit: deposit});
+
+        if (existingBid.bidder != address(0)) {
+            // If the bid exists, update the deposit
+            existingBid.deposit += deposit;
+        } else {
+            // If the bid does not exist, create a new one
+            poolEpochBids[id][_epoch][bidder] = newBid;
+        }
+
+        //Update poolEpochManager if deposit is largest
+        if (poolEpochManager[id][_epoch].deposit < deposit) {
+            poolEpochManager[id][_epoch] = newBid;
+        }
+
+        _updateEpochBids();
+    }
+
+    /// @inheritdoc IAmAmm
+    function cancelBid(PoolId id, uint40 _epoch) external virtual override {
+        /// -----------------------------------------------------------------------
+        /// Validation
+        /// -----------------------------------------------------------------------
+
+        address msgSender = LibMulticaller.senderOrSigner();
+
+        if (
+            !_amAmmEnabled(id) || poolEpochBids[id][_epoch][msgSender].deposit == 0
+                || _epoch <= _getEpoch(id, block.timestamp)
+        ) {
+            revert AmAmm__InvalidBid();
+        }
+
+        /// -----------------------------------------------------------------------
+        /// State updates
+        /// -----------------------------------------------------------------------
+
+        delete poolEpochBids[id][_epoch][msgSender];
+
+        Bid memory topBid = getHighestDepositBid(id, _epoch);
+        poolEpochManager[id][_epoch] = topBid;
+
+        _updateEpochBids();
+    }
+
+    /// @inheritdoc IAmAmm
+    function withdrawBid(PoolId id, uint40 _epoch, uint128 _amount) external virtual override {
```

Comment:
`poolEpochManager` doesn't need a deposit per epoch, as we could just have the hash table for all deposits. Hence this should be just a simple withdrawal from the `deposits[address]` mapping table.

Reply:
I'm not sure what you mean by this. In any case we would need the epoch as you could have bid in multiple epochs?